### PR TITLE
Supply a Git pre-commit hook for tools/check-typo

### PR DESCRIPTION
**Pull request by dra27**
This GPR provides a pre-commit script which can be used as a Githook to automate the use of `tools/check-typo` when developing.

I have ensured that `tools/check-typo` operates on the *index* and not the *working tree* (you can test this by adding a long line or a tab character to a file, doing `git add -u`, correcting the file in your working tree and then running `git commit`) but I haven't checked that the other sections of the script read the "correct" version of .gitattributes and so forth. This is something which should be addressed if this is generalised to the CI (which should really check each commit in turn on a pull request).

### CODE REVIEW

**Comment by dra27** [comment]
I know he'll have other things on his mind today and tomorrow, but this should definitely not be merged without @damiendoligez approving the changes in `tools/check-typo`.

**Comment by dra27** [comment]
Hmm - need to figure out why the copyright-header warning is triggering for all these scripts, but this is nearly there...

**Comment by dra27** [comment]
Lovely - so apparently gawk regexps don't work properly in Ubuntu 12.4.5.

This all appears to be working now...

**Comment by shindere** [comment]
@dra27 didn't have time to look into the code so far, sorry.

General question: would it be easy / doable to extract the code that
adds check-typo to travis and make a PR with just that? To me the two
features are distinct and it would simplify discussions to have them in
separate PRs.

**Comment by dra27** [comment]
@shindere - please see the caml-devel email I just sent! These are three stacked features - while we can merge any or all of them, they are related so I prefer to leave one GPR while we discuss and then split it before merging.

**Comment by dra27** [comment]
Incidentally, the bulk of the code is to do with Travis - the pre-commit hook only needs a tiny tweak in tools/check-typo

**Comment by shindere** [comment]
First of all, thanks for having rewritten the history, @dra27. It is
easier to read like this.

I have two technical questions.

1. What is the svnrules variable for, actually?
I realise it was there before, I am wondering whether it shouldn't be
renamed to something more appropriate?

2. On Travis, is it mandatory to create a check-typo-failed file?
Would it be possible to keep track of the failure in a shell variable or
do you really need something more persistent?

**Comment by dra27** [comment]
@shindere - thanks! I expect that all instances of svn could be renamed to git - I think it's just the rules that have been read from the .gitattributes file (I think the svn equivalent would have been properties)? I wrote the code defensively to allow for spaces in file names (because I'm a Windows user!) which means that the while loop is a subshell ... I don't think it's possible to communicate the state back without a file?

**Comment by shindere** [comment]
David Allsopp (2017/04/20 02:54 -0700):
> @shindere - thanks! I expect that all instances of svn could be
> renamed to git - I think it's just the rules that have been read from
> the .gitattributes file (I think the svn equivalent would have been
> properties)?

Couldn't we just _not_ mention any revision control system and use
"rules"?

> I wrote the code defensively to allow for spaces in file
> names (because I'm a Windows user!) which means that the while loop is
> a subshell ...

I just went quickly over the code, so sorry if my question is naïve.
I think it's nice to accept file names with spaces, but I do not
understand why this implies using a sub shell?

> I don't think it's possible to communicate the state
> back without a file?

Even by using different exit codes?

**Comment by damiendoligez** [comment]
@shindere we can't just call them "rules" because there are two kinds of rules, the default ones (rules) and the ones given by git attributes (svnrules).

@dra27 the changes to check-typo look OK to me (except I don't really understand the changes to the git commands) but the consensus on the developers list was that it's better to check the end result of a PR rather than every commit. I'm not sure how this relates to the fact that you're working on the index rather than the working tree.

**Comment by dra27** [comment]
@damiendoligez - the point from the dev meeting was addressed, so having rebased I think this is ready to go.

I will rebase & update #1288 **which should be merged shortly before this one**.

File: tools/ci/travis/travis-ci.sh
Comment (damiendoligez) [inline tools/ci/travis/travis-ci.sh]: Could you find a better name for this function? It doesn't necessarily check all commits anymore.

**Review by damiendoligez — APPROVED** [review]
Looks good, except for one remark.

File: tools/ci/travis/travis-ci.sh
Comment (dra27) [reply tools/ci/travis/travis-ci.sh]: Done - changed this one to `CheckTypo` and the previous function to `CheckTypoTree`, which better reflects what it's doing.

**Comment by dra27** [comment]
This is ready to go, but must wait until the repo fully passes `tools/check-typo` (I'm just testing an updated GPR which will then want splitting into bits)

**Comment by shindere** [comment]
David Allsopp (2018/06/08 07:24 -0700):
> This is ready to go, but must wait until the repo fully passes
> `tools/check-typo` (I'm just testing an updated GPR which will then
> want splitting into bits)

@dra27 I seem to remember you mentionned ocamltest will require some
changes, may I help with this?

**Comment by dra27** [comment]
@shindere - yes, indeed. I'm just running tests locally on a branch which includes everything and I'll then split it up into smaller branches for review (and because I think one or two might be wanted on the 4.07 testsuite).

**Comment by shindere** [comment]
David Allsopp (2018/06/08 07:31 -0700):
> @shindere - yes, indeed. I'm just running tests locally on a branch
> which includes everything and I'll then split it up into smaller
> branches for review (and because I think one or two might be wanted on
> the 4.07 testsuite).

Okay, many thanks for your hard work, @dra27.

I'll make sure to review at least the ocamltest-related bits.

**Comment by dra27** [comment]
Rebased and three improvements:

 - The pre-commit hook now uses the cached version of .gitattributes, rather than the working copy (so that means that if you push a file with white-at-eol on purpose, you must also commit the change in .gitattributes to add the attribute - previously you could leave the file by accident in your working directly)
 - The UTF-8 validation change in check-typo now works here too
 - Git has a nicer display of whitespace errors which the pre-commit hook was using. The only problem is that this didn't respect our `white-at-eol` and `white-at-eof` rules. It now does.

**Comment by dra27** [comment]
@shindere - in https://github.com/ocaml/ocaml/pull/1829#issuecomment-396155419 you came up with the great idea of running check-typo on Inria's CI... which is the best script to place that in?

**Comment by dra27** [comment]
OK - the changes made certainly warrant another review from someone, please ... I'm just running `check-typo` again, but it'd be good to get this in relatively soon so that we (hopefully) continue to comply with our linting tool!

**Comment by shindere** [comment]
David Allsopp (2018/06/14 07:14 -0700):
> @shindere - in
> https://github.com/ocaml/ocaml/pull/1829#issuecomment-396155419 you
> came up with the great idea of running check-typo on Inria's CI...
> which is the best script to place that in?

Perhaps `tools/ci/inria/extra-checks` which is run on only one machine.

But, was that such a great idea actually?

I mean, it's certainly nice to do that, but once the trunk will be
clean, won't we ant to catch the check-typo issues at the PR stage
rather than after merge?

Thanks!

**Comment by dra27** [comment]
I realised that we might want to update the githook in future and it might be useful for users to know that - the most recent commit compares the hook with the branch you're working on. I can't find a way to persuade `git diff` to compare a file in the repo with a physically different file on disk, so the hook now requires `bash`

**Comment by dra27** [comment]
@shindere - well, we do still push manually and also merge when Travis is red, so things may slip through!

**Comment by shindere** [comment]
David Allsopp (2018/06/14 14:56 +0000):
> @shindere - well, we do still push manually and also merge when Travis
> is red, so things may slip through!

Sure. I just meant that we could perhaps try to catch at least some lint
violations before merge, I realise it won't be enough.

**Comment by dra27** [comment]
Ah, sorry - this GPR already does that; there's a Travis job added (see https://travis-ci.org/ocaml/ocaml/jobs/392292527)

**Comment by shindere** [comment]
David Allsopp (2018/06/14 08:44 -0700):
> Ah, sorry - this GPR already does that; there's a Travis job added
> (see https://travis-ci.org/ocaml/ocaml/jobs/392292527)

Oh nice! And it tests only those files which are actually modified by
the PR, right?

**Comment by dra27** [comment]
Indeed, it only tests files which have been changed (or we'll get even crosser with how long Travis takes to complete!). It was a bit too much work to get `check-typo` just to check one's own changes and not issues which were already present in the file, hence the massive sweep of the repo to eliminate previous `check-typo` crimes!

**Comment by shindere** [comment]
David Allsopp (2018/06/14 08:47 -0700):
> Indeed, it only tests files which have been changed (or we'll get even
> crosser with how long Travis takes to complete!).

Sure!

> It was a bit too
> much work to get `check-typo` just to check one's own changes and not
> issues which were already present in the file, hence the massive sweep
> of the repo to eliminate previous `check-typo` crimes!

Yeah, I think it's not a big deal testing the whole changed files and
it's really great that the repo is correct check-typo-wise!!

That's one thing less to do at release time!

**Comment by dra27** [comment]
@shindere - does that look sufficient in https://github.com/ocaml/ocaml/pull/1148/commits/3d259e3239bc637a2eee0e12a2412553a97478cb? Are those tests only run on `trunk`?

**Comment by shindere** [comment]
David Allsopp (2018/06/15 01:17 -0700):
> @shindere - does that look sufficient in
> https://github.com/ocaml/ocaml/pull/1148/commits/3d259e3239bc637a2eee0e12a2412553a97478cb? Are those tests only run on `trunk`?

Hmm why do you need to test that the script is executable?

I think I would have written `./tools/check-typo` to run the command,
rather than just `tools/check-typo`, because as far as I know the
current directory won't be in the PATH.

**Comment by dra27** [comment]
@shindere - I was more testing that the script exists and, given that it's a script, `-x` seems better than `-e`? `./tools/check-typo` and `tools/check-typo` are equivalent - there's a slash in there, so it's never `PATH`-searched?!

**Comment by shindere** [comment]
David Allsopp (2018/06/15 01:43 -0700):
> @shindere - I was more testing that the script exists and, given that
> it's a script, `-x` seems better than `-e`?

But why do you need to test that it exists before running it?

In case it does not exist, wouldn't trying to run it triger an error
anyway?

> b`./tools/check-typo` and `tools/check-typo` are equivalent - there's
> a slash in there, so it's never `PATH`-searched?!

Ah I wasn't aware of that! Thanks!

**Comment by dra27** [comment]
@shindere - I was just going along with a neater error message than "command not found"!

**Comment by shindere** [comment]
David Allsopp (2018/06/15 02:24 -0700):
> @shindere - I was just going along with a neater error message than
> "command not found"!

Ah! Well, to be honest given that the script is part of the repo I find
this a bit overkill (I would prefer not having to do that systematically
for every script of the repo one owuld wnat to run...) but if you
like it that way, I'm fine.

**Comment by dra27** [comment]
@damiendoligez - I've made enough subsequent changes that I'd like to double-check that this still has your approval before merging?

**Comment by damiendoligez** [comment]
@dra27 yes, we're still good.

**Comment by dra27** [comment]
Right, several things fixed:
 - The hook script wasn't marked as executable (I developed it on Windows, so didn't notice...)
 - Files which have been deleted would result in an error message - they're now filtered out
 - The prune attribute wasn't taken into account, which meant that commits to the manual would fail. The process for checking that is more involved than would be nice. Ideally, we'd change the gitattributes to be `/manual/**` and then every file would show-up with prune however then the directory itself doesn't, so we'd need two entries for every pruned directory. So I opted for a recursive bash function because, well, why not.
 - Technically, the prune attribute should be being applied to directories in the root only (as it happens, `manual/manual` was also getting it), so I tidied that).

These issues were found because I've done a scan of check-typo over all open PRs which don't require rebasing (because of conflicts). In the interests of getting this merged, assuming it passes CI, I'll merge but if anyone wants to pick up any issues in the last 3 commits, please do and I'll open a new GPR.

**Pull request merged**
